### PR TITLE
8257513: C2: assert((constant_addr - _masm.code()->consts()->start()) == con.offset())

### PR DESCRIPTION
**Pull request by chhagedorn**
Running the testcase with `-XX:+StressCodeBuffers` initializes the code buffer with a small size:
https://github.com/openjdk/jdk/blob/0c8cc2cde47bed3d5c3edc203869068a0676812b/src/hotspot/share/opto/output.cpp#L1339-L1340
This leads to more frequent expansions in `ConstantTable::emit()` when calling the `_masm.*_constant()` methods. In the `T_VOID` case, we additionally fill the jump-table with dummy words:
https://github.com/openjdk/jdk/blob/0c8cc2cde47bed3d5c3edc203869068a0676812b/src/hotspot/share/opto/constantTable.cpp#L146-L150

When there is an expansion during this loop, then `constant_addr` is still pointing to an outdated address before expanding the code buffer and relocating the data. This lets the assertion fail when subtracting the new address `_masm.code()->consts()->start()` after the expansion.

I extended this assertion for `T_VOID` to take an expansion into account. 

There is a second issue that is currently not handled correctly. The `_masm.*_constant()` methods could return `NULL` when we failed to allocate new space in an expansion which, however, rarely happens. But when running with `-XX:+StressCodeBuffers`, `NULL` is much more likely to be returned because we intentionally free a blob from time to time:
https://github.com/openjdk/jdk/blob/0c8cc2cde47bed3d5c3edc203869068a0676812b/src/hotspot/share/asm/codeBuffer.cpp#L846-L854

I therefore suggest to also change the associated assertions in `ConstantTable::emit()` into bailout code to handle this case (as also done in the `LIR_Assembler` using the `AbstractAssembler::*_constant()` methods).

Thanks,
Christian

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8257513](https://bugs.openjdk.java.net/browse/JDK-8257513): C2: assert((constant_addr - _masm.code()->consts()->start()) == con.offset())

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1803/head:pull/1803`
`$ git checkout pull/1803`

### CODE REVIEW

File: test/hotspot/jtreg/compiler/codecache/TestStressCodeBuffers.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8257513
+ * @summary Stress testing code buffers resulted in an assertion failure due to not taking expand calls into account
+ *          which can fail more often with -XX:+StressCodeBuffers. Perform some more sanity flag testing.
+ * @run main/othervm -Xcomp -XX:-TieredCompilation -XX:+StressCodeBuffers compiler.codecache.TestStressCodeBuffers
+ * @run main/othervm -Xcomp -XX:+StressCodeBuffers compiler.codecache.TestStressCodeBuffers
```
Comment (vnkozlov) [inline test/hotspot/jtreg/compiler/codecache/TestStressCodeBuffers.java:30]: StressCodeBuffers is debug flag. You need to add `-XX:+IgnoreUnrecognizedVMOptions` flag or `@requires vm.debug == true`

**Review by vnkozlov — CHANGES_REQUESTED** [review]
Code changes seems fine.

File: test/hotspot/jtreg/compiler/codecache/TestStressCodeBuffers.java (hunk 1 above)
Comment (chhagedorn) [reply test/hotspot/jtreg/compiler/codecache/TestStressCodeBuffers.java:30]: Thanks, I fixed it.

**Review by vnkozlov — APPROVED** [review]
Good.

**Review by TobiHartmann — APPROVED** [review]
Looks good to me.

**Comment by chhagedorn** [comment]
Thank you Tobias for your review!

**Pull request closed without merging**